src/cache.jl

```diff
@@ -18,14 +17,13 @@ const specialization_counter = Ref{UInt}(0)
     length(mthds) == 1 || return (:(throw(MethodError(job.source.f,job.source.tt))))
     mtypes, msp, m = mthds[1]
     mi = ccall(:jl_specializations_get_linfo, Ref{MethodInstance}, (Any, Any, Any), m, mtypes, msp)
-    ci = retrieve_code_info(mi)
-    @assert isa(ci, CodeInfo)
+    ci = retrieve_code_info(mi)::CodeInfo
```

Comment:
Shouldn't we hit our own cache here? https://github.com/JuliaGPU/GPUCompiler.jl/blob/b610c12a7279a54d328f47907dffd9ff67d4c37a/src/jlgen.jl#L242

Which also has the lower-bound for the validity?  At least that should be our source of truth since we no longer care about what Base thinks is valid.

Reply:
We're executing in a generated function, and can't get the cache there (`ci_cache` takes an actual CompilerJob), let alone trigger inference if the method instance wasn't in our cache...

Reply:
I'm not sure why our use of `retrieve_code_info` here is guaranteed to return a CodeInfo (and not Nothing, as looking in our cache does since it's not popoulated -- this check is called _before_ compiling), but it's not terribly wrong, as we only use it for tracking redefinitions. That means we're missing out on redefinitions of device overrides though.

Reply:
Hm getting the AbsInt cache should be fine? I think instead of looking at the `mi`, you want to look at the `CodeInstance` since that has the world-bounds. Normally you could look at `mi->cache` to get the head of the cache, which is a `CodeInstance`, but since we are using an AbstractInterpreter we need to look at the associated cache there. 

Thinking more about this, you want an uninferred CI here right? 
So maybe calling https://github.com/JuliaLang/julia/blob/15c19c8a6b1c1a87e6dd24fd058de5192826b250/base/reflection.jl#L1042 is thre right direction? That at least uses the `CodeInstance` as metadata to set the `min_world` and `max_world`

Reply:
The codeinfo's world bounds are always -1/typemax, weirdly.